torch/csrc/distributed/c10d/ProcessGroupXCCL.cpp

```diff
@@ -0,0 +1,414 @@
+        break;
+    }
+  }
+}
+
+} // namespace
+
+static std::mutex xcclCommDevIdxMapMutex;
+static std::unordered_map<std::shared_ptr<xcclComm_t>, int> xcclCommDevIdxMap;
+constexpr int64_t kSynchronizeBusyWaitMillis = 10;
+
+ProcessGroupXCCL::WorkXCCL::WorkXCCL(
+    at::Device& device,
+    int rank,
+    OpType opType,
+    const std::optional<std::vector<at::Tensor>>& inputs)
+    : Work(rank, opType, "profilingTitle", inputs),
+      device_(device),
+      workStartTime_(std::chrono::steady_clock::now()) {
+  unsigned char enable_timing = 0;
+  xcclEndEvent_ = std::make_shared<at::xpu::XPUEvent>(enable_timing);
+}
+
+ProcessGroupXCCL::WorkXCCL::WorkXCCL(const WorkXCCL& w)
+    : Work(w.rank_, w.opType_),
+      device_(w.device_),
+      blockingWait_(w.blockingWait_),
+      workStartTime_(w.workStartTime_),
+      xcclEndEvent_(w.xcclEndEvent_) {}
+
+ProcessGroupXCCL::WorkXCCL::~WorkXCCL() = default;
+
+bool ProcessGroupXCCL::WorkXCCL::checkTimeout(
+    std::optional<std::chrono::milliseconds> timeout) {
+  auto currentTimepoint = std::chrono::steady_clock::now();
+  auto timeElapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
+      currentTimepoint - workStartTime_);
+  std::chrono::milliseconds opTimeout = std::chrono::milliseconds(60000);
+
+  auto workTimeout = timeout ? *timeout : opTimeout;
+
+  if (timeElapsed < workTimeout)
+    return false;
+  return true;
+}
+
+bool ProcessGroupXCCL::WorkXCCL::isCompleted() {
+  for (auto& ret : rets) {
+    bool flag;
+    try {
+      TORCH_CHECK(flag = ret.test());
+    } catch (...) {
+      finishAWorkXCCLError(std::current_exception());
+      return true;
+    }
+    if (!flag) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void ProcessGroupXCCL::WorkXCCL::synchronize() {
+  synchronizeInternal(kNoTimeout);
+}
+
+void ProcessGroupXCCL::WorkXCCL::synchronizeStream() {
+  auto currentStream = at::xpu::getCurrentXPUStream(device_.index());
+  // Block the current stream on the XCCL stream
+  xcclEndEvent_->block(currentStream);
+}
+
+void ProcessGroupXCCL::WorkXCCL::synchronizeInternal(
+    std::chrono::milliseconds timeout) {
+  synchronizeStream();
+
+  if (blockingWait_) {
+    while (!isCompleted()) {
+      bool timedOut = checkTimeout(
+          timeout == kNoTimeout ? std::nullopt : std::make_optional(timeout));
+      if (timedOut) {
+        break;
+      }
+      std::this_thread::sleep_for(
+          std::chrono::milliseconds(kSynchronizeBusyWaitMillis));
+    }
+  }
+}
+
+bool ProcessGroupXCCL::WorkXCCL::wait(std::chrono::milliseconds timeout) {
+  synchronizeInternal(timeout);
+  for (auto& event : rets) {
+    event.wait();
+  }
+  rets.clear();
+  return true;
+}
+
+ProcessGroupXCCL::ProcessGroupXCCL(
+    const c10::intrusive_ptr<Store>& store,
+    int rank,
+    int size)
+    : Backend(rank, size), store_(store) {
+  blockingWait_ = getCvarBool(TORCH_XCCL_BLOCKING_WAIT, false);
+  init();
+
+  {
```

Comment:
You need to check its launcher is IMPI or not. Only set CCL_LOCAL_RANK/CCL_LOCAL_SIZE for non-MPI launcher